src/hotspot/share/cds/classPrelinker.cpp

```diff
@@ -0,0 +1,224 @@
+#include "cds/archiveBuilder.hpp"
+#include "cds/classPrelinker.hpp"
+#include "classfile/systemDictionary.hpp"
+#include "classfile/vmClasses.hpp"
+#include "memory/resourceArea.hpp"
+#include "oops/constantPool.hpp"
+#include "oops/instanceKlass.hpp"
+#include "oops/klass.inline.hpp"
+#include "runtime/handles.inline.hpp"
+
+ClassPrelinker* ClassPrelinker::_singleton = NULL;
+
+bool ClassPrelinker::is_vm_class(InstanceKlass* ik) {
+  return (_vm_classes.get(ik) != NULL);
+}
+
+void ClassPrelinker::add_one_vm_class(InstanceKlass* ik) {
+  bool created;
+  _vm_classes.put_if_absent(ik, &created);
+  if (created) {
+    InstanceKlass* super = ik->java_super();
+    if (super != NULL) {
+      add_one_vm_class(super);
+    }
+    Array<InstanceKlass*>* ifs = ik->local_interfaces();
+    for (int i = 0; i < ifs->length(); i++) {
+      add_one_vm_class(ifs->at(i));
+    }
+  }
+}
+
+ClassPrelinker::ClassPrelinker() {
+  assert(_singleton == NULL, "must be");
+  _singleton = this;
+  for (auto id : EnumRange<vmClassID>{}) {
+    add_one_vm_class(vmClasses::klass_at(id));
+  }
+}
+
+ClassPrelinker::~ClassPrelinker() {
+  assert(_singleton == this, "must be");
+  _singleton = NULL;
+}
+
+bool ClassPrelinker::can_archive_resolved_vm_class(InstanceKlass* cp_holder, InstanceKlass* resolved_klass) {
+  if (!is_vm_class(resolved_klass)) {
+    return false;
+  }
+  if (!cp_holder->is_shared_boot_class() &&
+      !cp_holder->is_shared_platform_class() &&
+      !cp_holder->is_shared_app_class()) {
+    // Custom loaders are not guaranteed to resolve the vmClasses to the
+    // ones resolved by the boot loader.
+    return false;
+  }
+  if (cp_holder->class_loader_data() != resolved_klass->class_loader_data()) {
+    // If they are defined by different loaders, it's possible for resolved_klass
+    // to be already defined, but is not yet resolved in cp_holder->class_loader().
+
+    // TODO: this check can be removed if we preload the vmClasses into
+    // platform and app loaders during VM bootstrap.
+    return false;
+  }
+
+  return true;
+}
+
+bool ClassPrelinker::can_archive_resolved_klass(InstanceKlass* cp_holder, Klass* resolved_klass) {
+  assert(!is_in_archivebuilder_buffer(cp_holder), "sanity");
+  assert(!is_in_archivebuilder_buffer(resolved_klass), "sanity");
+
+  if (resolved_klass->is_instance_klass()) {
+    InstanceKlass* ik = InstanceKlass::cast(resolved_klass);
+    if (can_archive_resolved_vm_class(cp_holder, ik)) {
+      return true;
+    }
+    if (cp_holder->is_subtype_of(ik)) {
+      // All super types of ik will be resolved in ik->class_loader() before
+      // ik is defined in this loader, so it's safe to archive the resolved klass reference.
+      return true;
+    }
+
+    // TODO -- allow objArray classes, too
+  }
+
+  return false;
+}
+
+Klass* ClassPrelinker::get_resolved_klass_or_null(ConstantPool* cp, int cp_index) {
+  if (cp->tag_at(cp_index).is_klass()) {
+    CPKlassSlot kslot = cp->klass_slot_at(cp_index);
```

Comment:
This is another place that CPKlassSlot leaks out.  It would be nice if these were in constantPool.  Maybe this function belongs in the constant pool.

Reply:
Changed to use ConstantPool::resolved_klass_at() instead.